extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java

```diff
@@ -411,147 +539,199 @@ public Integer get() {
-                            columnWriter.addPage(bufferToWrite, nullValue, transferObject.rowCount());
-                        } else {
-                            columnWriter.addPageNoNulls(bufferToWrite, transferObject.rowCount());
+    }
+
+    private static <DATA_TYPE> boolean tryEncodeDictionary(final ColumnDefinition<DATA_TYPE> columnDefinition,
+            final ParquetInstructions writeInstructions,
+            final RowSet rowSet,
+            final ColumnSource<DATA_TYPE> columnSource,
+            final ColumnSource<?> lengthSource,
+            final RowSet lengthRowSet,
+            final int targetRowsPerPage,
+            final LongSupplier rowStepGetter,
+            final LongSupplier valuesStepGetter,
+            final int pageCount,
+            final ColumnWriter columnWriter) throws IOException {
+        // Note: We only support strings as dictionary pages. Knowing that, we can make some assumptions about chunk
+        // types and avoid a bunch of lambda and virtual method invocations. If we decide to support more, than
+        // these assumptions will need to be revisited.
+        Assert.eqTrue(supportsDictionary(columnSource.getType()), "ColumnSource supports dictionary");
+
+        final boolean useDictionaryHint = writeInstructions.useDictionary(columnDefinition.getName());
+        final int maxKeys = useDictionaryHint ? Integer.MAX_VALUE : writeInstructions.getMaximumDictionaryKeys();
+        try {
+            final List<IntBuffer> pageBuffers = new ArrayList<>();
+            Binary[] encodedKeys = new Binary[Math.min(INITIAL_DICTIONARY_SIZE, maxKeys)];
+
+            final TObjectIntHashMap<String> keyToPos =
+                    new TObjectIntHashMap<>(Constants.DEFAULT_CAPACITY, Constants.DEFAULT_LOAD_FACTOR, -1);
+            int keyCount = 0;
+            boolean hasNulls = false;
+            try (final ChunkSource.GetContext context = columnSource.makeGetContext(targetRowsPerPage);
+                    final RowSequence.Iterator it = rowSet.getRowSequenceIterator()) {
+                for (int curPage = 0; curPage < pageCount; curPage++) {
+                    final RowSequence rs = it.getNextRowSequenceWithLength(valuesStepGetter.getAsLong());
+                    // noinspection unchecked
+                    final ObjectChunk<String, Values> chunk =
+                            (ObjectChunk<String, Values>) columnSource.getChunk(context, rs);
+                    final IntBuffer posInDictionary = IntBuffer.allocate((int) rs.size());
+                    for (int vi = 0; vi < chunk.size(); ++vi) {
+                        final String key = chunk.get(vi);
+                        int dictionaryPos = keyToPos.get(key);
+                        if (dictionaryPos == keyToPos.getNoEntryValue()) {
+                            if (key == null) {
+                                hasNulls = true;
+                                dictionaryPos = QueryConstants.NULL_INT;
```

Comment:
I thought it was more 'correct' to use the thing from the class.  Not that it makes any difference.

Reply:
That's totally correct. I'm just saying you can skip the `dictionaryPos = QueryConstants.NULL_INT;` assignment if you use `QueryConstants.NULL_INT` instead of `-1` as the no-entry value.